### PR TITLE
[BE][export] add data-dependent section to export tutorial

### PR DESCRIPTION
**Pull request by pianpwk**
Adds brief section, with pointer to more in-depth data-dependent errors doc

## Checklist

- [ ] The issue that is being fixed is referred in the description (see above "Fixes #ISSUE_NUMBER")
- [ ] Only one issue is addressed in this pull request
- [x] Labels from the issue that this PR is fixing are added to this pull request
- [x] No unnecessary issues are included into this pull request.

cc @williamwen42 @msaroufim @anijain2305

### CODE REVIEW

File: intermediate_source/torch_export_tutorial.py [hunk 1]
```diff
@@ -629,6 +629,188 @@ def forward(self, x, y):
...
 }
 
+######################################################################
+# Data-dependent errors
+# ---------------------
+#
+# While trying to export models, you have may have encountered errors like ``Could not guard on data-dependent expression`` or ``Could not extract specialized integer from data-dependent expression``.
+# Obscure as they may seem, the reasoning behind their existence, and their resolution, is actually quite straightforward.
+#
+# These errors exist because ``torch.export()`` compiles programs using ``FakeTensors``, which symbolically represent their real tensor counterparts (e.g. they may have the same or equivalent symbolic properties
```
Comment (svekars) [inline intermediate_source/torch_export_tutorial.py:639]:
```suggestion
# These errors exist because ``torch.export()`` compiles programs using ``FakeTensors``, which symbolically represent their real tensor counterparts (for example, they may have the same or equivalent symbolic properties
```

File: intermediate_source/torch_export_tutorial.py [hunk 2]
```diff
@@ -629,6 +629,188 @@ def forward(self, x, y):
...
+# - sizes, strides, dtypes, etc.), but diverge in one major respect: they do not contain any data values. While this avoids unnecessary memory usage and expensive computation, it does mean that the compiler may
```
Comment (svekars) [inline intermediate_source/torch_export_tutorial.py:640]:
```suggestion
# - sizes, strides, dtypes, and so on), but diverge in one major respect: they do not contain any data values. While this avoids unnecessary memory usage and expensive computation, it does mean that the compiler may
```

File: intermediate_source/torch_export_tutorial.py [hunk 3]
```diff
@@ -629,6 +629,188 @@ def forward(self, x, y):
...
+# FakeTensor tracing isn't providing the information required.
+#
+# Let's talk about where data-dependent values appear in programs. Common sources are calls like ``item()``, ``tolist()``, or ``torch.unbind()`` that extract scalar values from tensors.
+# How are these values represented in the exported program? In the ``Constraints/Dynamic Shapes`` section, we talked about allocating symbols to represent dynamic input dimensions, and the same happens here -
+# we allocate symbols for every data-dependent value that appears in the program. The important distinction is that these are "unbacked" symbols or "unbacked SymInts", in contrast to the "backed" symbols/SymInts
+# allocated for input dimensions. The "backed/unbacked" nomenclature refers to the presence, or absence, of a "hint" for the symbol: a concrete value backing the symbol, that can inform the compiler how to proceed.
+#
+# For dynamic input shapes (backed SymInts), these hints are taken from the shapes of the sample inputs provided, which explains why sample input shapes direct the compiler in control-flow branching.
```
Comment (svekars) [inline intermediate_source/torch_export_tutorial.py:649]:
```suggestion
# For dynamic input shapes (backed ``SymInts``), these hints are taken from the shapes of the sample inputs provided, which explains why sample input shapes direct the compiler in control-flow branching.
```

File: intermediate_source/torch_export_tutorial.py [hunk 4]
```diff
@@ -629,6 +629,188 @@ def forward(self, x, y):
...
+# - sizes, strides, dtypes, etc.), but diverge in one major respect: they do not contain any data values. While this avoids unnecessary memory usage and expensive computation, it does mean that the compiler may
+# struggle with user code that relies on data values. In short, if the compiler requires a concrete, specialized value that is dependent on tensor data in order to proceed, it will error, complaining that
+# FakeTensor tracing isn't providing the information required.
+#
+# Let's talk about where data-dependent values appear in programs. Common sources are calls like ``item()``, ``tolist()``, or ``torch.unbind()`` that extract scalar values from tensors.
+# How are these values represented in the exported program? In the ``Constraints/Dynamic Shapes`` section, we talked about allocating symbols to represent dynamic input dimensions, and the same happens here -
+# we allocate symbols for every data-dependent value that appears in the program. The important distinction is that these are "unbacked" symbols or "unbacked SymInts", in contrast to the "backed" symbols/SymInts
```
Comment (svekars) [inline intermediate_source/torch_export_tutorial.py:646]:
```suggestion
# we allocate symbols for every data-dependent value that appears in the program. The important distinction is that these are "unbacked" symbols or "unbacked" ``SymInts``, in contrast to the "backed" symbols/``SymInts``
```

File: intermediate_source/torch_export_tutorial.py [hunk 5]
```diff
@@ -629,6 +629,188 @@ def forward(self, x, y):
...
+# allocated for input dimensions. The "backed/unbacked" nomenclature refers to the presence, or absence, of a "hint" for the symbol: a concrete value backing the symbol, that can inform the compiler how to proceed.
+#
+# For dynamic input shapes (backed SymInts), these hints are taken from the shapes of the sample inputs provided, which explains why sample input shapes direct the compiler in control-flow branching.
+# On the other hand, data-dependent values are derived from FakeTensors during tracing, and by default lack hints to inform the compiler, hence the name "unbacked symbols/SymInts".
```
Comment (svekars) [inline intermediate_source/torch_export_tutorial.py:650]:
```suggestion
# On the other hand, data-dependent values are derived from ``FakeTensors`` during tracing, and by default lack hints to inform the compiler, hence the name "unbacked symbols" or ``SymInts``.
```

**Review by svekars — COMMENTED** [review]
a couple of editorial nits.

**Review by yushangdi — APPROVED** [review]
LGTM!

Thanks for the nice tutorial!

File: intermediate_source/torch_export_tutorial.py [hunk 6]
```diff
@@ -629,6 +629,190 @@ def forward(self, x, y):
     "bool_val": None,
 }
 
+######################################################################
+# Data-dependent errors
+# ---------------------
+#
+# While trying to export models, you have may have encountered errors like "Could not guard on data-dependent expression", or "Could not extract specialized integer from data-dependent expression".
+# These errors exist because ``torch.export()`` compiles programs using FakeTensors, which symbolically represent their real tensor counterparts. For example, they may have equivalent symbolic properties
```
Comment (avikchaudhuri) [inline intermediate_source/torch_export_tutorial.py:637]: Cut "For example, ...may"

You can rephrase by highlighting the main difference first, or use "While they have equivalent..., they diverge in that..."

File: intermediate_source/torch_export_tutorial.py [hunk 7]
```diff
@@ -629,6 +629,190 @@ def forward(self, x, y):
...
+# (e.g. sizes, strides, dtypes), but diverge in that FakeTensors do not contain any data values. While this avoids unnecessary memory usage and expensive computation, it does mean that export may struggle
```
Comment (avikchaudhuri) [inline intermediate_source/torch_export_tutorial.py:638]: reword "may struggle"

File: intermediate_source/torch_export_tutorial.py [hunk 8]
```diff
@@ -629,6 +629,190 @@ def forward(self, x, y):
...
+# with parts of user code where compilation relies on data values. In short, if the compiler requires a concrete, data-dependent value in order to proceed, it will error out, complaining that
+# FakeTensor tracing isn't providing the information required.
```
Comment (avikchaudhuri) [inline intermediate_source/torch_export_tutorial.py:640]: Generally speaking, you should highlight data-dependence as the main cause rather than making it about fake tensors. Also, export programming model has sections on shape vs. data dependence.

File: intermediate_source/torch_export_tutorial.py [hunk 9]
```diff
@@ -629,6 +629,190 @@ def forward(self, x, y):
...
+#
+# Data-dependent values appear in many places, and common sources are calls like ``item()``, ``tolist()``, or ``torch.unbind()`` that extract scalar values from tensors.
+# How are these values represented in the exported program? In the `Constraints/Dynamic Shapes <https://pytorch.org/tutorials/intermediate/torch_export_tutorial.html#constraints-dynamic-shapes>`_
+# section, we talked about allocating symbols to represent dynamic input dimensions.
+# The same happens here: we allocate symbols for every data-dependent value that appears in the program. The important distinction is that these are "unbacked" symbols or "unbacked SymInts",
+# in contrast to the "backed" symbols/SymInts allocated for input dimensions. The "backed/unbacked" nomenclature refers to the presence/absence of a "hint" for the symbol:
```
Comment (avikchaudhuri) [inline intermediate_source/torch_export_tutorial.py:646]: symbols / SymInts vs. symbols or SymInts, choose one

Comment (avikchaudhuri) [reply intermediate_source/torch_export_tutorial.py:646]: Again, the export programming model explains what backed vs. unbacked means, so maybe link there.

File: intermediate_source/torch_export_tutorial.py [hunk 10]
```diff
@@ -629,6 +629,190 @@ def forward(self, x, y):
...
+    torch.tensor([2, 3]),
+)
+ep = export(Foo(), inps)
+print(ep)
+
+######################################################################
+# The result is that 3 unbacked symbols (notice they're prefixed with "u", instead of the usual "s" for input shape/backed symbols) are allocated and returned:
+# 1 for the ``item()`` call, and 1 for each of the elements of ``y`` with the ``tolist()`` call.
+# Note from the range constraints field that these take on ranges of ``[-int_oo, int_oo]``, not the default ``[0, int_oo]`` range allocated to input shape symbols,
+# since we literally have no information on what these values are - they don't represent sizes, so don't necessarily have positive values.
```
Comment (avikchaudhuri) [inline intermediate_source/torch_export_tutorial.py:671]: cut "literally"

File: intermediate_source/torch_export_tutorial.py [hunk 11]
```diff
@@ -629,6 +629,190 @@ def forward(self, x, y):
...
+        if a // 2 >= 5:
+            return y + 2
+        else:
+            return y * 5
+
+######################################################################
+# Here we actually need the "hint", or the concrete value of ``a`` for the compiler to decide whether to trace ``return y + 2`` or ``return y * 5`` as the output.
+# Because we trace with FakeTensors, we don't know what ``a // 2 >= 5`` actually evaluates to, and export errors out with "Could not guard on data-dependent expression ``u0 // 2 >= 5 (unhinted)``".
+#
+# So how do we actually export this? Unlike ``torch.compile()``, export requires full graph compilation, and we can't just graph break on this. Here's some basic options:
```
Comment (avikchaudhuri) [inline intermediate_source/torch_export_tutorial.py:692]: cut "actually"

Comment (avikchaudhuri) [reply intermediate_source/torch_export_tutorial.py:692]: "this" what? maybe "this code"

Comment (avikchaudhuri) [reply intermediate_source/torch_export_tutorial.py:692]: Here *are* some options:

**Review by avikchaudhuri — APPROVED** [review]
Please address nits
